Add "Recommended Plugins" input box

mikachan opened the pull request:
This PR adds a "Recommended Plugins" input to the wp-admin page and editor versions of the plugin.

Closes https://github.com/WordPress/create-block-theme/issues/399.

### Testing Instructions:
1. Open the wp-admin page or the editor sidebar panel
2. Add some text to the recommended plugins input box
3. Export the theme
4. The text should be added to the theme's readme, under a "Recommended Plugins" heading

### Screenshots:

wp-admin:
<img width="827" alt="image" src="https://github.com/WordPress/create-block-theme/assets/1645628/b71e520f-f39a-4249-9758-e373373a015e">

Editor sidebar panel:
<img width="280" alt="image" src="https://github.com/WordPress/create-block-theme/assets/1645628/77511a29-c25d-4daa-9112-9e497bccdbd5">

mikachan committed:
Add Recommended Plugins input

mikachan committed:
Add Recommended Plugins section to theme readme

mikachan committed:
Hide Recommended Plugins input on blank theme option

mikachan committed:
Merge branch 'trunk' into add/recommended-plugin-input

mikachan committed:
Update recommended plugins link

pbking commented:
I see the additional form on the editor.  However, nothing sticks when I edit it.
<img width="305" alt="image" src="https://github.com/WordPress/create-block-theme/assets/146530/cbdefde7-d905-4a50-8854-8df5795ed86c">

I don't see the form on the wp-admin 'Create Blank' form.  (Though I do see it on `Clone` and `Create Child`).

The forms don't seem to add the value to the exported theme though.

mikachan committed:
Refactor readme logic

mikachan committed:
Fix file name reference in comment

mikachan committed:
Add recommended_plugins to React state

mikachan committed:
Handle recommended_plugins in API endpoints

mikachan committed:
Generate readme file on theme update

mikachan committed:
Handle readme file on theme update

mikachan committed:
Increase spacing before image credits

mikachan committed:
Move image credits to end of readme

mikachan committed:
Add additional check before running recommended_plugins_section

mikachan committed:
Remove recommended plugins input from create panel

mikachan commented:
Thanks for testing, @pbking!

> I see the additional form on the editor. However, nothing sticks when I edit it.

It looks like the editor version was only updating the style.css file, so I've updated this PR so that the readme.txt file is also updated from the editor.

> I don't see the form on the wp-admin 'Create Blank' form.

This is intentional (similar to how the image credits input works), as I was thinking that people wouldn't want to add recommended plugins to a blank theme. I've also removed the 'Recommended Plugins' input from the 'Create' sidebar panel, as I'm now thinking it's only useful on the 'Update' panel for now. We could revisit this when we align the wp-admin version and editor versions closer together.

mikachan commented on admin/create-theme/theme-readme.php:
```diff
@@ -4,37 +4,35 @@ class Theme_Readme {
 		$original_theme_credits = $new_copyright_section ? self::original_theme_credits( $name, $is_parent_theme ) : '';
+		$copyright_section      = self::copyright_section( $new_copyright_section, $original_theme_credits, $name, $copy_year, $author, $image_credits );
 
-		$default_copyright_section = "== Copyright ==
```
I've refactored this and moved it lower down so that it's contained in `copyright_section()`.

mikachan committed:
Create separate update_readme_txt function

mikachan committed:
Include updates to description, author and WP version in readme file on update

vcanales reviewed (commented):
Could we add a placeholder with an example? It's been hard to find an explanation of this section; for instance, it's not listed in the docs https://developer.wordpress.org/themes/basics/main-stylesheet-style-css/

I wondered if I should input a URL or if just the plugin's name would suffice, for example. I can also picture folks wondering if they should separate plugins with a comma or if the textarea will be printed into the document as is.

pbking committed:
Adding readme file data to theme metadatda API

mikachan committed:
Default recommendedPlugins to empty string

mikachan committed:
Use actual error message in failure response

mikachan committed:
Remove extra line breaks from plugins section

jffng commented on src/editor-sidebar/update-panel.js:
```diff
@@ -32,10 +33,26 @@ export const UpdateThemePanel = () => {
+				return response.data;
+			}
+			return {
+				recommendedPlugins: '',
```
Why do we need to return an object with only this key as an empty string? Seems like a strange exception.

mikachan replied:
I've updated this so it's now returning an empty object with no specific keys.

mikachan committed:
Use correct success message

jffng reviewed (commented):
I tested and verified this is working as expected in the wp-admin and site editor flows.

While some of the changes are just formatting, I do think this is a lot of code to add for what is effectively just reading/writing some unstructured text to the readme, without enforcement or validation of those plugins coming from wp.org, for example. Unless I am missing something? It's a fair suggestion, but what's the real utility? For example, Gutenberg has a mechanism to prompt users to install respective plugins for blocks that are missing from templates / patterns a theme supplies.

I don't mean to block the PR, just adding my general thoughts.

jffng reviewed (approved):
On second thought, I can see how having this info is useful for theme authors to inform users what plugins to install, and a lot of the changes are just refactoring / formatting, so giving it the ✅ .

mikachan committed:
Move getThemeReadmeData into a useEffect

Fixes uncontrolled input React warning

mikachan committed:
Default to returning empty object

mikachan committed:
Add comment

mikachan committed:
Fix extra line breaks being added to readme

mikachan committed:
Add placeholders to Recommended Plugin inputs

mikachan commented:
Thanks, both, for testing this out!

> It's been hard to find an explanation of this section

I was also looking for an example, and the best reference I can find is here: https://make.wordpress.org/themes/handbook/review/required/#6-plugins. I think a placeholder is a good idea, I've added one here: https://github.com/WordPress/create-block-theme/pull/402/commits/1b293ab6db297e2e874eebe3f8abbab480bf6503, which looks like this:

```
Plugin Name
https://wordpress.org/plugins/plugin-name/
Plugin Description
```

> and a lot of the changes are just refactoring / formatting

Yes... sorry about that, this snowballed as I saw loads of improvements to the existing readme logic! Thank you for working through it.

@pbking and I paired on this earlier today, and with our latest changes, the 'Recommended Plugins' textarea in the editor will now pre-populate with data from the current readme file, if this section already exists. This functionality also makes it easier to expand in the future for other parts of the readme file (e.g. image credits). I think this enhances the UX, as the user sees some feedback from updating the theme data.

vcanales reviewed (approved):
Thanks for adding the placeholder! I gave it another go and it LGTM.

pbking committed:
Small tweak to regex when replacing recommended plugins section to ensure other sections are not modified

pbking commented:
I made a teeny tiny tweak to the regex processing to ensure that if another section is in readme.txt following the recommended plugins section that the title `== TITLE HERE ==` of the section remains unchanged (it was removing the first `==`'s)

Otherwise this cleaned up great!  Well done @mikachan !  This was a blast to pear on!
